docs/tech_specs/0005_eink_widget_endpoint.md

```diff
@@ -0,0 +1,216 @@
+  if (widgetJson) widgetJson = Object.values(widgetJson)[0]
+
+  return widgetJson ? <Widget data={widgetJson} /> : null
+};
+```
+
+And then suddenly, the widget page is rendered! Mercury will be taking screenshots of these pages and sending the image to the screen, only to update with a new screenshot if the data for that widget changes.
+
+Example 1: Footer. POST to http://localhost:4000/v2/screen/gl_eink_v2/widget with body
+```
+{
+    "widget": {
+        "footer": {
+            "mode_cost": "$2.40",
+            "mode_icon": "subway-negative-black.svg",
+            "mode_text": "Subway",
+            "text": "For real-time predictions and fare purchase locations:",
+            "type": "fare_info_footer",
+            "url": "mbta.com/stops/place-bcnwa"
+        }
+    }
+}
+```
+
+displays:
+![example eink footer][example eink footer]
+
+Example 2: Partial alert. POST to http://localhost:4000/v2/screen/gl_eink_v2/widget with body
+```
+{
+    "widget": {
+        "medium": {
+            "body": "Shuttle buses replacing Green Line E branch service",
+            "header": "Shuttle Buses",
+            "icon": "bus",
+            "route_pills": [
+                {
+                    "color": "green",
+                    "text": "Green Line E",
+                    "type": "text"
+                }
+            ],
+            "type": "alert",
+            "url": "mbta.com/alerts"
+        }
+    }
+}
+```
+
+displays:
+![example partial alert][example eink partial alert]
+
+Example 3: Takeover alert. POST to http://localhost:4000/v2/screen/gl_eink_v2/widget with body
+```
+{
+    "widget": {
+        "body": {
+            "full_body_bottom_screen": {
+                "type": "bottom_screen_filler"
+            },
+            "full_body_top_screen": {
+                "body": "Heath Street closed",
+                "header": "Station Closed",
+                "icon": "x",
+                "route_pills": [
+                    {
+                        "color": "green",
+                        "text": "Green Line E",
+                        "type": "text"
+                    }
+                ],
+                "type": "full_body_alert",
+                "url": "mbta.com/alerts"
+            },
+            "type": "body_takeover"
+        }
+    }
+}
+```
+
+displays:
+![example takeover alert][example eink takeover alert]
+
+# Unresolved Questions
+- Are the MappingContext and ResponseMapperContext needed as wrappers for `<WidgetPage />`?
+- What about slots?
+  - Since Mercury is building their own frontend skin, I believe the layout and sizing needs to be managed on their end. Is that true?
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+???
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Are there any alternatives that would avoid needing to re-run the backend code? (And are there any drawbacks of the current approach that make this re-run worth it?)
```

Comment:
How about this as a possible alternative?

What if Mercury kept long-running instances of each of our e-ink clients running on their side, and instead of passing widget data back to us for rendering, they just passed it to their dedicated client, which would render it for screenshotting?

(The client would still need this new `/widget` route so that it would know to render one individual widget)

That way,
1. We would not need to handle any additional requests after each data request
2. We would not need to send up the entire GL/bus e-ink react app in response to each of these widget requests (we wouldn't receive these requests at all)
3. We would just need to periodically send the react app up to their thing, to keep it up to date.

Reply:
Hmmmmm I'm not sure what you mean by "their dedicated client". What client will they have where our react app is running? They're building out their own frontend app for our data

Reply:
I mean client as "thing that uses the data that our server responds with". In the current proposal, we already have them loading and running our react app somewhere, ephemerally, in order to render the single-widget view that they then screenshot. There's some machine or process dedicated to that.

This would just change things so that the client app doesn't get loaded anew whenever any long-lifespanned widget's data changes. They would keep one long-lived client running, and send widget data to it—via a websocket or something—repeatedly for rendering.

They would send their "load a widget page and screenshot it" requests over there instead, and our server would only need to send new html/JS to them when we deploy changes to the client code.

Again this is an alternative but... it's almost certainly not a _better_ alternative.

Reply:
I just remembered that you had mentioned Mercury prefers the originally proposed approach and already uses it with their other transit customers—that's reason enough to discard this alternative. Resolving!